File: lib/imagereader/fdiimagereader.cc
#include "lib/core/globals.h"
#include "lib/config/config.h"
#include "lib/config/flags.h"
#include "lib/data/sector.h"
#include "lib/imagereader/imagereader.h"
#include "lib/data/image.h"
#include "lib/config/proto.h"
#include "lib/core/logger.h"
#include "lib/config/config.pb.h"
#include <algorithm>
#include <iostream>
#include <fstream>

// reader based on this partial documentation of the FDI format:
// https://www.pc98.org/project/doc/hdi.html

class FdiImageReader : public ImageReader
{
public:
    FdiImageReader(const ImageReaderProto& config): ImageReader(config) {}

    std::unique_ptr<Image> readImage() override
    {
        std::ifstream inputFile(
            _config.filename(), std::ios::in | std::ios::binary);
        if (!inputFile.is_open())
            error("cannot open input file");

        Bytes header(32);
        inputFile.read((char*)header.begin(), header.size());
        ByteReader headerReader(header);
        if (headerReader.seek(0).read_le32() != 0)
            error("FDI: could not find FDI header, is this a FDI file?");

        // we currently don't use fddType but it could be used to automatically
        // select profile parameters in the future
        //
        int fddType = headerReader.seek(4).read_le32();
        int headerSize = headerReader.seek(0x08).read_le32();
        int sectorSize = headerReader.seek(0x10).read_le32();
        int sectorsPerTrack = headerReader.seek(0x14).read_le32();
        int sides = headerReader.seek(0x18).read_le32();
        int tracks = headerReader.seek(0x1c).read_le32();

        inputFile.seekg(headerSize);

        std::unique_ptr<Image> image(new Image);
        int trackCount = 0;
        for (int track = 0; track < tracks; track++)
        {
            if (inputFile.eof())
                break;

            for (int side = 0; side < sides; side++)
            {
                std::vector<unsigned> sectors;
                for (int i = 0; i < sectorsPerTrack; i++)
                    sectors.push_back(i + 1);

                for (int sectorId : sectors)
                {
                    Bytes data(sectorSize);
                    inputFile.read((char*)data.begin(), data.size());

                    const auto& sector = image->put(track, side, sectorId);
                    sector->status = Sector::OK;
                    sector->data = data;
                }
            }

            trackCount++;
        }

        auto layout = _extraConfig.mutable_layout();
        if (globalConfig()->encoder().format_case() ==
            EncoderProto::FormatCase::FORMAT_NOT_SET)
        {
            auto ibm = _extraConfig.mutable_encoder()->mutable_ibm();
            auto trackdata = ibm->add_trackdata();
            trackdata->set_target_clock_period_us(2);

            auto layoutdata = layout->add_layoutdata();
            auto physical = layoutdata->mutable_physical();
            switch (fddType)
            {
                case 0x90:
                    log("FDI: automatically setting format to 1.2MB (1024 byte "
                        "sectors)");
                    trackdata->set_target_rotational_period_ms(167);
                    layoutdata->set_sector_size(1024);
                    for (int i = 0; i < 9; i++)
                        physical->add_sector(i);
                    break;

                case 0x30:
                    log("FDI: automatically setting format to 1.44MB");
                    trackdata->set_target_rotational_period_ms(200);
                    layoutdata->set_sector_size(512);
                    for (int i = 0; i < 18; i++)
                        physical->add_sector(i);
                    break;

                default:
                    error(
                        "FDI: unknown fdd type 0x{:2x}, could not determine "
                        "write profile automatically",
                        fddType);
                    break;
            }
        }

        image->calculateSize();
        const Geometry& geometry = image->getGeometry();
        log("FDI: read {} tracks, {} sides, {} kB total",
            geometry.numTracks,
            geometry.numSides,
            ((int)inputFile.tellg() - headerSize) / 1024);

        layout->set_tracks(geometry.numTracks);
        layout->set_sides(geometry.numSides);

        return image;
    }
};

std::unique_ptr<ImageReader> ImageReader::createFdiImageReader(
    const ImageReaderProto& config)
{
    return std::unique_ptr<ImageReader>(new FdiImageReader(config));
}

File: java/com/cowlark/fluxengine/imagereader/FdiImageReader.java
package com.cowlark.fluxengine.imagereader;

import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.core.Logger;
import com.cowlark.fluxengine.data.Geometry;
import com.cowlark.fluxengine.data.Image;
import com.cowlark.fluxengine.data.Sector;
import com.cowlark.fluxengine.encoders.EncoderProto;
import com.cowlark.fluxengine.ibm.IbmEncoderProto;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/* Reader based on this partial documentation of the FDI format:
 * https://www.pc98.org/project/doc/hdi.html
 */
public class FdiImageReader extends ImageReader
{
    public FdiImageReader(ImageReaderProto config, ConfigProto fullConfig)
    {
        super(config, fullConfig);
    }

    @Override
    public Image readImage()
    {
        Bytes data;
        try
        {
            data = new Bytes(Files.readAllBytes(Path.of(config.getFilename())));
        } catch (IOException e)
        {
            throw new FluxEngineException("cannot open input file");
        }

        ByteReader headerReader = new ByteReader(data.slice(0, 32));
        if (headerReader.seek(0).readLe32() != 0)
            throw new FluxEngineException("FDI: could not find FDI header, is this a FDI file?");

        /* we currently don't use fddType but it could be used to automatically
         * select profile parameters in the future */
        int fddType = headerReader.seek(4).readLe32();
        int headerSize = headerReader.seek(0x08).readLe32();
        int sectorSize = headerReader.seek(0x10).readLe32();
        int sectorsPerTrack = headerReader.seek(0x14).readLe32();
        int sides = headerReader.seek(0x18).readLe32();
        int tracks = headerReader.seek(0x1c).readLe32();

        ByteReader br = new ByteReader(data.slice(headerSize));

        Image image = new Image();
        int trackCount = 0;
        for (int track = 0; track < tracks; track++)
        {
            if (br.eof())
                break;

            for (int side = 0; side < sides; side++)
            {
                for (int sectorId = 1; sectorId <= sectorsPerTrack; sectorId++)
                {
                    Bytes sectorData = br.read(sectorSize);

                    Sector sector = image.put(track, side, sectorId);
                    sector.status = Sector.Status.OK;
                    sector.data = sectorData;
                }
            }

            trackCount++;
        }

        ConfigProto.Builder extra = ConfigProto.newBuilder();
        com.cowlark.fluxengine.config.LayoutProto.Builder layout = extra.getLayoutBuilder();
        if (fullConfig.getEncoder().getFormatCase() == EncoderProto.FormatCase.FORMAT_NOT_SET)
        {
            IbmEncoderProto.Builder ibm = extra.getEncoderBuilder().getIbmBuilder();
            IbmEncoderProto.TrackdataProto.Builder trackdata = ibm.addTrackdataBuilder();
            trackdata.setTargetClockPeriodUs(2);

            com.cowlark.fluxengine.config.LayoutProto.LayoutdataProto.Builder layoutdata =
                    layout.addLayoutdataBuilder();
            com.cowlark.fluxengine.config.SectorListProto.Builder physical =
                    layoutdata.getPhysicalBuilder();
            switch (fddType)
            {
                case 0x90:
                    Logger.logf("FDI: automatically setting format to 1.2MB (1024 byte sectors)");
                    trackdata.setTargetRotationalPeriodMs(167);
                    layoutdata.setSectorSize(1024);
                    for (int i = 0; i < 9; i++)
                        physical.addSector(i);
                    break;

                case 0x30:
                    Logger.logf("FDI: automatically setting format to 1.44MB");
                    trackdata.setTargetRotationalPeriodMs(200);
                    layoutdata.setSectorSize(512);
                    for (int i = 0; i < 18; i++)
                        physical.addSector(i);
                    break;

                default:
                    throw new FluxEngineException(String.format(
                            "FDI: unknown fdd type 0x%02x, could not determine write " +
                                    "profile automatically", fddType));
            }
        }

        image.calculateSize();
        Geometry geometry = image.getGeometry();
        Logger.logf("FDI: read " + geometry.numCylinders + " tracks, " + geometry.numHeads +
                " sides, " + (data.size() - headerSize) / 1024 + " kB total");

        layout.setTracks(geometry.numCylinders);
        layout.setSides(geometry.numHeads);

        extraConfig = extra.build();
        return image;
    }
}
